File: plugin/src/main/java/com/github/shadowsocks/plugin/PluginOptions.java
package com.github.shadowsocks.plugin;

import android.text.TextUtils;

import java.util.HashMap;
import java.util.Objects;
import java.util.StringTokenizer;

/**
 * Helper class for processing plugin options.
 *
 * Based on: https://github.com/apache/ant/blob/588ce1f/src/main/org/apache/tools/ant/types/Commandline.java
 *
 * @author Mygod
 */
public final class PluginOptions extends HashMap<String, String> {
    public PluginOptions() {
        super();
    }
    public PluginOptions(int initialCapacity) {
        super(initialCapacity);
    }
    public PluginOptions(int initialCapacity, float loadFactor) {
        super(initialCapacity, loadFactor);
    }

    // TODO: this method is not needed since API 24
    public String getOrDefault(Object key, String defaultValue) {
        String v;
        return (((v = get(key)) != null) || containsKey(key))
                ? v
                : defaultValue;
    }

    private PluginOptions(String options, boolean parseId) {
        this();
        if (TextUtils.isEmpty(options)) return;
        final StringTokenizer tokenizer = new StringTokenizer(options + ';', "\\=;", true);
        final StringBuilder current = new StringBuilder();
        String key = null;
        while (tokenizer.hasMoreTokens()) {
            String nextToken = tokenizer.nextToken();
            if ("\\".equals(nextToken)) current.append(tokenizer.nextToken());
            else if ("=".equals(nextToken) && key == null) {
                key = current.toString();
                current.setLength(0);
            } else if (";".equals(nextToken)) {
                if (key != null) {
                    put(key, current.toString());
                    key = null;
                } else if (current.length() > 0)
                    if (parseId) id = current.toString(); else put(current.toString(), null);
                current.setLength(0);
                parseId = false;
            } else current.append(nextToken);
        }
    }
    public PluginOptions(String options) {
        this(options, true);
    }
    public PluginOptions(String id, String options) {
        this(options, false);
        this.id = id;
    }

    public String id = "";

    private static void append(StringBuilder result, String str) {
        for (int i = 0; i < str.length(); ++i) {
            char ch = str.charAt(i);
            switch (ch) {
                case '\\': case '=': case ';': result.append('\\'); // intentionally no break
                default: result.append(ch);
            }
        }
    }
    public String toString(boolean trimId) {
        final StringBuilder result = new StringBuilder();
        if (!trimId) if (TextUtils.isEmpty(id)) return ""; else append(result, id);
        for (Entry<String, String> entry : entrySet()) {
            if (result.length() > 0) result.append(';');
            append(result, entry.getKey());
            String value = entry.getValue();
            if (value != null) {
                result.append('=');
                append(result, value);
            }
        }
        return result.toString();
    }
    @Override
    public String toString() {
        return toString(true);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        if (!super.equals(o)) return false;
        PluginOptions that = (PluginOptions) o;
        return Objects.equals(id, that.id) && super.equals(that);
    }

    @Override
    public int hashCode() {
        return Objects.hash(super.hashCode(), id);
    }
}

File: plugin/src/main/java/com/github/shadowsocks/plugin/PluginOptions.kt
/*******************************************************************************
 *                                                                             *
 *  Copyright (C) 2017 by Max Lv <max.c.lv@gmail.com>                          *
 *  Copyright (C) 2017 by Mygod Studio <contact-shadowsocks-android@mygod.be>  *
 *                                                                             *
 *  This program is free software: you can redistribute it and/or modify       *
 *  it under the terms of the GNU General Public License as published by       *
 *  the Free Software Foundation, either version 3 of the License, or          *
 *  (at your option) any later version.                                        *
 *                                                                             *
 *  This program is distributed in the hope that it will be useful,            *
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of             *
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the              *
 *  GNU General Public License for more details.                               *
 *                                                                             *
 *  You should have received a copy of the GNU General Public License          *
 *  along with this program. If not, see <http://www.gnu.org/licenses/>.       *
 *                                                                             *
 *******************************************************************************/

package com.github.shadowsocks.plugin

import java.util.*

/**
 * Helper class for processing plugin options.
 *
 * Based on: https://github.com/apache/ant/blob/588ce1f/src/main/org/apache/tools/ant/types/Commandline.java
 */
class PluginOptions : HashMap<String, String?> {
    var id = ""

    constructor() : super()
    constructor(initialCapacity: Int) : super(initialCapacity)
    constructor(initialCapacity: Int, loadFactor: Float) : super(initialCapacity, loadFactor)

    private constructor(options: String?, parseId: Boolean) : this() {
        @Suppress("NAME_SHADOWING")
        var parseId = parseId
        if (options.isNullOrEmpty()) return
        val tokenizer = StringTokenizer(options + ';', "\\=;", true)
        val current = StringBuilder()
        var key: String? = null
        while (tokenizer.hasMoreTokens()) {
            val nextToken = tokenizer.nextToken()
            when (nextToken) {
                "\\" -> current.append(tokenizer.nextToken())
                "=" -> if (key == null) {
                    key = current.toString()
                    current.setLength(0)
                } else current.append(nextToken)
                ";" -> {
                    if (key != null) {
                        put(key, current.toString())
                        key = null
                    } else if (current.isNotEmpty())
                        if (parseId) id = current.toString() else put(current.toString(), null)
                    current.setLength(0)
                    parseId = false
                }
                else -> current.append(nextToken)
            }
        }
    }

    constructor(options: String?) : this(options, true)
    constructor(id: String, options: String?) : this(options, false) {
        this.id = id
    }

    private fun append(result: StringBuilder, str: String) = (0 until str.length)
            .map { str[it] }
            .forEach {
                when (it) {
                    '\\', '=', ';' -> {
                        result.append('\\') // intentionally no break
                        result.append(it)
                    }
                    else -> result.append(it)
                }
            }

    fun toString(trimId: Boolean): String {
        val result = StringBuilder()
        if (!trimId) if (id.isEmpty()) return "" else append(result, id)
        for ((key, value) in entries) {
            if (result.isNotEmpty()) result.append(';')
            append(result, key)
            if (value != null) {
                result.append('=')
                append(result, value)
            }
        }
        return result.toString()
    }

    override fun toString(): String = toString(true)

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        return javaClass == other?.javaClass && super.equals(other) && id == (other as PluginOptions).id
    }
    override fun hashCode(): Int = Objects.hash(super.hashCode(), id)
}
